### PR TITLE
Media 2020 queries

### PR DESCRIPTION
**Pull request by smatei**
adding sql queries for [2020 media](https://github.com/HTTPArchive/almanac.httparchive.org/issues/900)

Using custom metrics:
- [x] srcset w/wo sizes
- [x] srcset x/w descriptor
- [x] srcset average candidates
- [x] picture formats
- [x] picture orientation
- [x] picture using min resolution
- [x] video autoplay muted
- [x] video preload
- [x] usage of alt text (from _markup custom metrics)

Without custom metrics:
- [ ]  images lazy loading (img having data-src with/without src)
- [x] Use of rel=preconnect for key 3rd party requests
- [x] images data-uri vs src attribute
- [x] images mimetype vs extension
- [x] _score_progressive_jpeg
- [x] video src and video with <source> child elements as well as the exensions in the <source> elements
- [x] <figure> and <figcaption> - ie. how many pages have <figure> and how many of those pages have <figure> with a <figcaption> as a child
- [x] for video player check if video.js is imported when a page has a <video> element or not
- [ ] Natural width/client width/height for all images
- [ ] for svg - check for large width/height SVGs vs small filesizes

### CODE REVIEW

**Comment by rviscomi** [comment]
Thanks @smatei! Please set this to "Draft" mode if it's not yet ready for review.

**Comment by rviscomi** [comment]
Hi @smatei just checking in on the status of this PR. How's it going?

**Comment by rviscomi** [comment]
@smatei Please ping this PR when it's ready for review.

**Comment by rviscomi** [comment]
@smatei could you update this PR description with a checklist of metrics that need to be and are implemented? That would help us keep track of your progress and see how close it is to being ready for review. We need to get all analysis finished no later than **this week** to unblock authors so they can get their chapters drafted in time for launch.

**Comment by smatei** [comment]
> 
> 
> @smatei could you update this PR description with a checklist of metrics that need to be and are implemented? That would help us keep track of your progress and see how close it is to being ready for review. We need to get all analysis finished no later than **this week** to unblock authors so they can get their chapters drafted in time for launch.

I will update this PR description today. I have to implement sqls that need to parse with regexp the response body. I will try to merge all the needed metrics in a single sql, so that it will be cheaper to run. I will try to break the results later in the analysis sheet.

**Comment by bazzadp** [comment]
> I will try to merge all the needed metrics in a single sql, so that it will be cheaper to run. I will try to break the results later in the analysis sheet.

Please don’t do this. It makes the SQL
quite complex to view - both for reviewers and also for readers who want to run or view  individual queries themselves. If running low on credits then reach out to @rviscomi who can sort you out to run them.

**Comment by rviscomi** [comment]
@smatei how many response body metrics do you have to query? Trying to get a sense of how expensive it would be as one monolithic query versus individual ones. Since I assume these queries will be limited to firstHtml, they should be much cheaper than a full response body scan.

Also I realize I forgot to provide you with sample data. I'll do that now.

**Comment by rviscomi** [comment]
@smatei here's the sample table: https://pantheon.corp.google.com/bigquery?p=httparchive&d=sample_data&t=summary_response_bodies_firstHtml&page=table

Also @bazzadp for context Stefan and I worked it out and I put him on the HA billing account to complete the rest of his queries. It's not WAI to need to query the response bodies and we spoke about custom metrics being the appropriate place to analyze that data, but there wasn't sufficient time to get those in. We need to be sure to implement those in time for next year (which is what we said last year when this happened 😬).

**Comment by smatei** [comment]
> 
> 
> @smatei here's the sample table: https://console.cloud.google.com/bigquery?p=httparchive&d=sample_data&t=summary_response_bodies_firstHtml&page=table
> 
> Also @bazzadp for context Stefan and I worked it out and I put him on the HA billing account to complete the rest of his queries. It's not WAI to need to query the response bodies and we spoke about custom metrics being the appropriate place to analyze that data, but there wasn't sufficient time to get those in. We need to be sure to implement those in time for next year (which is what we said last year when this happened 😬).

Thank you for the summary table. The custom metrics were squeezed in a week or so at the end of July, merged in a hurry, and set to be run in August. A better solution for next year would be to allow one more month for implementing custom metrics, let's say August, and run them on September. That would save us a lot of processing power.

**Comment by smatei** [comment]
> 
> 
> @smatei how many response body metrics do you have to query? Trying to get a sense of how expensive it would be as one monolithic query versus individual ones. Since I assume these queries will be limited to firstHtml, they should be much cheaper than a full response body scan.
> 
> Also I realize I forgot to provide you with sample data. I'll do that now.

The `httparchive.almanac.summary_response_bodies` with date = '2020-08-01' AND firstHtml would be 25TB. That would result in $125 per query. For 10 queries that would be over $1000.

**Comment by rviscomi** [comment]
@smatei be aware that BigQuery's cost estimates don't take clustered fields into account. So as long as you have a `WHERE firstHtml` clause on the `summary_response_bodies` query it should only actually process ~1.2 TB.

image

**Comment by rviscomi** [comment]
Hi @smatei do you think you can have this PR ready for review by the end of the day? We need to hand this off to the authors to start the draft ASAP.

**Comment by rviscomi** [comment]
@smatei @bazzadp how's this coming along?

**Comment by smatei** [comment]
> 
> 
> @smatei @bazzadp how's this coming along? Perhaps it's worth opening this PR up for review to get the completed queries out of the way, and reopen a new PR with the remaining ones.

I had a zoom meeting on Thursday with the media team, we discussed most of the remaining points from the chapter outline. A lot of them were not achievable from the database point of view (for example, count a videos with/without an audio channel - calling an external script was necessary for that). There are some metrics left, I am working on them. But I believe it would be good to review the existing scripts and add the rest of them on another PR. 

Why does the Lint fail with `Unable to process command '::set-env name=VALIDATE_ALL_CODEBASE::false' successfully.`?

**Comment by bazzadp** [comment]
> Why does the Lint fail with `Unable to process command '::set-env name=VALIDATE_ALL_CODEBASE::false' successfully.`?

GitHub changed something here. Can you merge the latest `main` into this branch and you’ll pick up the new linter run files and this should go.

After that remove this from Draft and I’ll review the queries today.

Let me know if you need a hand with any of that.

**Comment by smatei** [comment]
> 
> 
> > Why does the Lint fail with `Unable to process command '::set-env name=VALIDATE_ALL_CODEBASE::false' successfully.`?
> 
> GitHub changed something here. Can you merge the latest `main` into this branch and you’ll pick up the new linter run files and this should go.
> 
> After that remove this from Draft and I’ll review the queries today.
> 
> Let me know if you need a hand with any of that.

For the main merge, please confirm that these are the commands:

`git fetch`
`git rebase origin/main`
`git push` (or should I use `git push --force-with-lease`?)

**Comment by bazzadp** [comment]
I normally [do these when working on the main repo](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Git-Guide#keeping-your-branch-up-to-date):

```
git fetch origin main
git merge main
git commit -a
git push
```

**Comment by smatei** [comment]
> 
> 
> I normally [do these when working on the main repo](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Git-Guide#keeping-your-branch-up-to-date):
> 
> ```
> git fetch origin main
> git merge main
> git commit -a
> git push
> ```

I've got an "already up to date" message. 

image

I have changed the status to "ready for review".

File: sql/2020/05_Media/image_mimetype_ext.sql
Comment (bazzadp) [inline sql/2020/05_Media/image_mimetype_ext.sql:7]: Again would be good to see as percentage. What do you think of this:

```suggestion
  client,
  ext,
  mimetype,
  SUM(COUNT(0)) OVER (PARTITION BY client) AS total_images,
  COUNT(0) AS ext_mime_image_count,
  SAFE_DIVIDE(COUNT(0), SUM(COUNT(0)) OVER (PARTITION BY client)) AS total_image_pct,
  SAFE_DIVIDE(COUNT(0), SUM(COUNT(0)) OVER (PARTITION BY client, ext)) AS ext_pct,
  SAFE_DIVIDE(COUNT(0), SUM(COUNT(0)) OVER (PARTITION BY client, mimetype)) AS mime_pct
```

File: sql/2020/05_Media/image_mimetype_ext.sql
Comment (bazzadp) [inline sql/2020/05_Media/image_mimetype_ext.sql:16]: `summary_response_bodies` only contains text responses to save space. So you basically only get SVG images fo this. Use `requests` which has same columns you need but for all requests:

```suggestion
    `httparchive.almanac.requests`
```

File: sql/2020/05_Media/image_mimetype_ext.sql
Comment (bazzadp) [inline sql/2020/05_Media/image_mimetype_ext.sql:26]: Incorrect column name and some nits about capitalisation and newline:

```suggestion
ORDER BY
  COUNT(0) desc;

```

File: sql/2020/05_Media/image_srcset_candidates.sql
Comment (bazzadp) [inline sql/2020/05_Media/image_srcset_candidates.sql:32]: Same comments as for `image_alt.sql` above. Should we measure both pages (i.e. at least one result) and total images?

File: sql/2020/05_Media/image_srcset_descriptor.sql
Comment (bazzadp) [inline sql/2020/05_Media/image_srcset_descriptor.sql:30]: Same again. I won't repeat this feedback for the rest of the queries but most of them look to be similar.

**Review by bazzadp — CHANGES_REQUESTED** [review]
A few suggestions. The feedback about using SUMs as well as COUNTIFs is something that applies to most of the queries so would like your opinion on that?

File: sql/2020/05_Media/big_non_custom_metrics.sql
Comment (smatei) [reply sql/2020/05_Media/big_non_custom_metrics.sql:20]: I did not run this yet, only tested on sample data. Anyway if we decide to get last year sqls and run them now we will end up spending a lot too.

File: sql/2020/05_Media/image_mimetype_ext.sql
Comment (smatei) [reply sql/2020/05_Media/image_mimetype_ext.sql]: You are right, it looked a little bit to strange to me that the SVG images were at the top. Now that I have changed to requests I have jpg, png, gif on top, but a lot of low count rows like this

image

26k rows. The idea of this, was to figure out if the mimetype is matching the extension, how could I limit this for each of the clients (desktop, mobile)? I can add them all to the spreadsheet, but I think that would be irrelevant.

File: sql/2020/05_Media/big_non_custom_metrics.sql
Comment (smatei) [reply sql/2020/05_Media/big_non_custom_metrics.sql:20]: By the way, Rick replied here https://github.com/HTTPArchive/almanac.httparchive.org/pull/1327#issuecomment-724097296 that it might not cost $127 per query as I suspected, but much less.

Comment (bazzadp) [reply sql/2020/05_Media/big_non_custom_metrics.sql:20]: Ah that's true. Still if we've got them cheaper let's use them. Haven't had a chance to dig into the custom metrics yet to confirm but hopefully will later today.

File: sql/2020/05_Media/image_mimetype_ext.sql
Comment (smatei) [reply sql/2020/05_Media/image_mimetype_ext.sql]: For now, I will get 100 rows for each client and dump them in the spreadsheet in case someone looks for this data.

Comment (bazzadp) [reply sql/2020/05_Media/image_mimetype_ext.sql]: There's always some complete gibberish so let's set a minimum size. The following gives 189 rows. Note I've also removed the grouping sub query as think useful to look at all images a site sends back.

```sql
#standardSQL
# images mimetype vs extension
SELECT
  client,
  ext,
  mimetype,
  COUNT(0) AS ext_mime_image_count,
  SUM(COUNT(0)) OVER (PARTITION BY client) AS total_images,
  SAFE_DIVIDE(COUNT(0), SUM(COUNT(0)) OVER (PARTITION BY client)) AS total_image_pct,
  SAFE_DIVIDE(COUNT(0), SUM(COUNT(0)) OVER (PARTITION BY client, ext)) AS ext_pct,
  SAFE_DIVIDE(COUNT(0), SUM(COUNT(0)) OVER (PARTITION BY client, mimetype)) AS mime_pct
FROM
  `httparchive.almanac.requests`
WHERE
  date = '2020-08-01' AND
  type='image'
GROUP BY
  client, ext, mimetype
HAVING
  ext_mime_image_count > 10000
ORDER BY
  ext_mime_image_count desc, client
```

Comment (bazzadp) [reply sql/2020/05_Media/image_mimetype_ext.sql]: > For now, I will get 100 rows for each client and dump them in the spreadsheet in case someone looks for this data.

Paths crossed. Yeah that's basically what I did. Tried 1000, then when saw that gave 400 rows, I tried 10,000. The other option is to set a LIMIT but difficult to do that for each CLIENT and also I personally prefer setting a hard limit.

File: sql/2020/05_Media/big_non_custom_metrics.sql
Comment (bazzadp) [reply sql/2020/05_Media/big_non_custom_metrics.sql:20]: OK looks like `` data is not available in a custom metric so I was wrong about that.

Resolving this. Though think we should look at SUM as well as COUNTIFs as per other feedback and need to change to the real table before merging.

File: sql/2020/05_Media/image_alt.sql
Comment (smatei) [reply sql/2020/05_Media/image_alt.sql:33]: I was thinking of percentages too. You are right, we should display both pages and instances, and let them choose. I will check all the sqls for the same pattern. At the end please check the names of the metrics. I will try to follow this guideline ("pages_with_img_pct", "imgs_alt_missing_pct").

File: sql/2020/05_Media/big_non_custom_metrics.sql
Comment (bazzadp) [inline sql/2020/05_Media/big_non_custom_metrics.sql:16]: We should leave them as between 0 and 1, rather than 0 and 100 as Google Sheets will handle the percentage. Note [this is different to have we used to do it in 2019](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Analysts'-Guide#formatting-results)

```suggestion
  COUNTIF(has_img_data_uri) / COUNTIF(has_img_src) AS pages_with_img_data_uri_pct,
  COUNTIF(rel_preconnect) / COUNT(0) AS rel_preconnect,
  COUNTIF(has_video_src) / COUNT(0) AS pages_with_video_src_pct,
  COUNTIF(has_video_source) / COUNT(0) AS pages_with_video_source_pct,
  COUNTIF(has_figure) / COUNT(0) AS pages_with_figure_pct,
  COUNTIF(has_figcaption) / COUNT(0) AS pages_with_figcaption_pct
```

Can you do that for all the queries?

File: sql/2020/05_Media/video_source_types.sql
Comment (bazzadp) [inline sql/2020/05_Media/video_source_types.sql:21]: I'm not convinced about this query. This will only pick up the first result, when a `<video>` element [can contain multiple sources](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source), such as:

```html
<video controls
    width="250"
    height="200"
    muted>
    <source src="/media/cc0-videos/flower.webm"
            type="video/webm">
    <source src="/media/cc0-videos/flower.mp4"
            type="video/mp4">
    This browser does not support the HTML5 video element.
</video>
```

What about below query to look at all the `types` (which doesn't have to be specified but should be to avoid a download for an unsupported type):

```sql
SELECT
  client,
  video_type,
  COUNT(0) AS video_type_count,
  SAFE_DIVIDE(COUNT(0), SUM(COUNT(0)) OVER (PARTITION BY client)) AS pct_video_types
FROM
    `httparchive.sample_data.summary_response_bodies_firstHtml`,
    #`httparchive.almanac.summary_response_bodies`,
    UNNEST(REGEXP_EXTRACT_ALL(body, '(?i)(<video.*?</video>)')) AS video,
    UNNEST(REGEXP_EXTRACT_ALL(video, r'(?i)type\s*=\s*["\'](video/[^\'";]*)')) AS video_type
WHERE
  date = '2020-08-01' AND
  firstHtml
GROUP BY
  client,
  video_type
ORDER BY
  client,
  video_type_count desc
```

This was basically stolen from [this query from 2019 media chapter](https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2019/04_Media/04_19.sql) but improved upon (at least I think so!)

**Comment by rviscomi** [comment]
@smatei I've pushed a commit to reformat all of the queries, mostly trivial capitalization and spacing fixes.

Note that I changed `big_non_custom_metrics.sql` to use case-insensitive regular expressions, which will be more accurate but may affect the results slightly.

I'm going to merge these changes as-is because we're launching in less than 2 weeks and need to get the analysis done ASAP. We can review any SQL changes in a follow-up PR as needed.

**Pull request merged**
